sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala

```diff
@@ -345,6 +348,28 @@ private[sql] class ParquetRelation2(
     // Schema of the whole table, including partition columns.
     var schema: StructType = _
 
+    def filterDataStatusesWithoutSummaries(
+      leaves: Seq[FileStatus],
+      dataStatuses: Seq[FileStatus]): Seq[FileStatus] = {
+
+      // Get the partitions that have summary files
+      val typeInference = sqlContext.conf.partitionColumnTypeInferenceEnabled()
+      val summariesPaths = metadataStatuses.map(_.getPath.getParent()) ++
+        commonMetadataStatuses.map(_.getPath.getParent())
+      val summariesPartitions = PartitioningUtils.parsePartitions(summariesPaths,
+        PartitioningUtils.DEFAULT_PARTITION_NAME, typeInference).partitions.toSet
+
+      dataStatuses.filterNot { d =>
+        val part = PartitioningUtils.parsePartitions(Seq(d.getPath.getParent()),
+          PartitioningUtils.DEFAULT_PARTITION_NAME, typeInference)
```

Comment:
Parsing partitions over a single path is dangerous. For example, consider a partition column whose values are hex number strings:

```
/base/p=01
/base/p=0a
```

When dealing with the first path only, the column type is inferred to be `IntegerType`. But when you take both paths into consideration, the column type turns to `StringType` instead.